### ISSUE TITLE
Provide a public method to change the selection of ui5-select programmatically

### ISSUE THREAD

**Issue by vladitasev** [SAP/ui5-webcomponents#1041]
Discuss and provide a suitable public API to change selection.

We have a read-only `selectedOption` property:

```js
	/**
	 * Currently selected option
	 * @readonly
	 * @type { ui5-option }
	 * @public
	 */
	get selectedOption() {
		return this.options.find(option => option.selected);
	}
```

What one must currently do is something like:

```js
[...$0.children].forEach(option => option.selected = option.textContent === "Compact")
```

We need an official API for this.

**Comment by tnurmi82** [comment]
I hope to see this improvement soon. I'd also need to change selection programmatically.

**Comment by Smux** [comment]
Yes, please. It's not a good approach right now :(

**Comment by ignipae** [comment]
Hi,
I would also also like to see a suitable public API.

BR
Julian

**Comment by vladitasev** [comment]
Hello @tnurmi82  @Smux @ignipae 

We've stayed away from this so long, because generally in HTML everything needs to be declarative

```html
	<ui5-select>
		<ui5-option>Cozy</ui5-option>
		<ui5-option selected>Compact</ui5-option>
		<ui5-option>Condensed</ui5-option>
	</ui5-select>
```

so if you bind `selected` and change your model, in whichever framework, the selected option will be changed automatically.

As far as I understand, you need an imperative way of doing the same.

In the above PR we're currently evaluating one possibility: namely to add a public setter

```js
const option1 = mySelect.options[1];
mySelect.selectedOption = option1;
```

This assumes you have a reference to the option though. Is that what you need, or rather setting by text, or something else?

Regards,
Vladi

**Comment by tnurmi82** [comment]
Hello @vladitasev , sorry for delayed response.

In my case (a React app) I have an UID for each option, set in attribute `data-uid`. The "currently valid" UID value is stored in state. When loading/rendering UI I need to automatically set correct option, so that user can see it. 
(The actual visible text value is language specific, could be in english or something else.) 

My solution was a small function which loops through all the options, checks the `data-uid` value, and does `option.setAttribute('selected', '');` for the desired option, and `option.removeAttribute('selected');` for all the others.

So in my case something like `mySelect.setSelectedOptionByAttributeValue(attrName, attrValue);` would have been most awesome, but setting by text would have also helped.

**Comment by Smux** [comment]
@vladitasev yes Setting the Option by using the object reference should be fine as well but for sure it would be better if setting it by text would also work :)

**Comment by PetyaMarkovaBogdanova** [comment]
A backlog item has been created on the subject and the team will start working on it accordingly.

**Comment by nnaydenow** [comment]
Internal reference: BGSOFUIPIRIN-5342

**Comment by vladitasev** [comment]
This feature is nice-to-have. The most universal solution would be to create a `setSelectedOption` public method, which gets a callback as the only argument. This callback is executed for all options and it selects the first/last (ideally only) option that it matches, depending on how you choose to implement it.

Usage:
```js
select.setSelectedOption(option => option.textContent === "some text");
```
This will select the option that has "some text" as textContent.

Example 2:
```js
select.setSelectedOption(option => option.getAttribute("data-country-attr") === "bg");
```
This will select the option with a certain data attribute

Example 3:
```js
select.setSelectedOption(option => option === someOption);
```
This will select an option by reference (if you have the reference already).

We could create an alternative syntax for this:
```js
select.setSelectedOption(someOption);
```

In summary, my proposal would be:
```
setSelectedOption(callback|object)
```
where you can pass the instance of the option that you want selected (if you already have it), or you may pass a callback that will be executed against each option and the first/last matching will be selected. All others will be deselected.

**Comment by ilhan007** [comment]
We had a discussion on the topic  and decided the declarative way of selecting an option is enough and recommended.

We considered providing a setter, a method to select an option as:
```js
select.setSelectedOption(option => option.textContent === "Compact");
```

But, the effort is not much different from what is currently possible
```js
[...$0.children].forEach(option => option.selected = option.textContent === "Compact")
```

That's why for time being, we don't plan to introduce such an API.
